### PR TITLE
Adapt for Python3.6

### LINKED ISSUE #2160: Python 3.6 support
This issue is for tracking support for Python 3.6.

### PR DESCRIPTION
**Pull request by sklam**
Closes #2160
Closes #2142 (supersedes)

### CODE REVIEW

**Comment by sklam** [comment]
Still need to implement the following opcodes:

* [ ] BUILD_TUPLE_UNPACK_WITH_CALL
* [ ] BUILD_CONST_KEY_MAP

**Comment by seibert** [comment]
Also looks like the refactoring of the bytecodes broke Python 2.7.

**Comment by sklam** [comment]
This now works with Python 3.6.0b3.  But we have no CI for it yet.

**Comment by seibert** [comment]
How do we want to handle the CI?  Setting up the Python 3.6 environment from conda-forge beta packages is still pretty wonky.  It will be easier once the release is official.  Maybe we want to merge this PR and then create an issue to remind us to add Python 3.6 to the CI when it is official?

**Comment by sklam** [comment]
I would prefer not to make a CI setup with the beta packages.   I don't see any harm for merging it early.  It would give us more time to exercise the new code paths.

File: numba/bytecode.py [hunk 1]
```diff
@@ -35,128 +35,6 @@ def get_code_object(obj):
     return getattr(obj, '__code__', getattr(obj, 'func_code', None))
 
 
-def _make_bytecode_table():
```
Comment (sklam) [inline numba/bytecode.py]: This table is no longer needed.   Originally, this table encodes the length of each instruction.  Python3.6 switched to fixed-length instruction word (16-bit).  It turns out pre-3.6 instruction length can be determined by whether the opcode is >= than `dis.HAVE_ARGUMENT`, which indicates the bytecode has 2-extra bytes for the argument.

Comment (pitrou) [inline numba/bytecode.py]: Nice simplification :-)

File: numba/dataflow.py [hunk 2]
```diff
@@ -181,7 +181,8 @@ def op_BUILD_MAP(self, info, inst):
 
     def op_BUILD_SET(self, info, inst):
         count = inst.arg
-        items = [info.pop() for _ in range(count)]
+        # Note: related python bug http://bugs.python.org/issue26020
+        items = list(reversed([info.pop() for _ in range(count)]))
```
Comment (pitrou) [inline numba/dataflow.py]: Can also be written as `[info.pop() for _ in range(count)][::-1]` (just saying :-))

**Comment by pitrou** [comment]
This looks overall good to me.

> I would prefer not to make a CI setup with the beta packages. I don't see any harm for merging it early. It would give us more time to exercise the new code paths.

Agreed.

**Pull request merged**